libc-top-half/musl/src/unistd/getcwd.c

```diff
@@ -14,12 +16,21 @@ char *getcwd(char *buf, size_t size)
 		errno = EINVAL;
 		return 0;
 	}
+#ifdef __wasilibc_unmodified_upstream // WASI libc uses a custom getcwd
 	long ret = syscall(SYS_getcwd, buf, size);
 	if (ret < 0)
 		return 0;
 	if (ret == 0 || buf[0] != '/') {
 		errno = ENOENT;
 		return 0;
 	}
+#else
+        size_t len = strlen(__wasilibc_cwd);
+        if (size < len + 1) {
+                errno = ERANGE;
+                return 0;
+        }
+        memcpy(buf, __wasilibc_cwd, len + 1);
+#endif
```

Comment:
In the case where `buf` is `NULL`, we should just `malloc(len + 1)` and `memcpy` into it, rather than artificially limiting it to `PATH_MAX` bytes. A key differences here is that musl is written for Linux where it has to allocate the buffer before making the syscall. For WASI, we can just do the `strlen` right away, check the length if using a pre-allocated buffer, and `malloc` if not.

So I think it makes to have our own `getcwd.c` in libc-bottom-half/sources/getcwd.c rather than base this on musl's getcwd.c.